virtcontainers/device/drivers/vhost_user_fs.go

```diff
@@ -0,0 +1,72 @@
+// Copyright (C) 2018 Red Hat, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package drivers
+
+import (
+	"encoding/hex"
+
+	"github.com/kata-containers/runtime/virtcontainers/device/api"
+	"github.com/kata-containers/runtime/virtcontainers/device/config"
+	"github.com/kata-containers/runtime/virtcontainers/utils"
+)
+
+// VhostUserFSDevice is a virtio-fs vhost-user device
+type VhostUserFSDevice struct {
+	*GenericDevice
+	config.VhostUserDeviceAttrs
+}
+
+// Device interface
+
+func (device *VhostUserFSDevice) Attach(devReceiver api.DeviceReceiver) (err error) {
+	skip, err := device.bumpAttachCount(true)
+	if err != nil {
+		return err
+	}
+	if skip {
+		return nil
+	}
+
+	// generate a unique ID to be used for hypervisor commandline fields
+	randBytes, err := utils.GenerateRandomBytes(8)
+	if err != nil {
+		return err
+	}
+	id := hex.EncodeToString(randBytes)
+
+	device.DevID = id
+	device.Type = device.DeviceType()
+
+	defer func() {
+		if err == nil {
+			device.AttachCount = 1
```

Comment:
I'm rather confused by this: the function calls `device.bumpAttachCount(true)` to increase `AttachCount` and then explicitly sets `AttachCount = 1` if no error. I see the same logic in reverse in `Detach()`. It seems therefore that the only valid `AttachCount` values are `0` and `1` but can't that be checked for rather than overriding what `bumpAttachCount()` is doing?

Maybe a few comments explaining the intent would help here?

Reply:
@jodh-intel I think I am the one to be blamed :crying_cat_face: 

The logic is confusing as you said but it's right. 
* `AttachCount == 0`: do real attach, and set AttachCound = 1;
* `AttachCount==1`: AttachCount++, but skip the real attact.
* `AttachCount==IntMax`: error

I'll create a new PR to improve the readability.

Reply:
Here it is: https://github.com/kata-containers/runtime/issues/1392

It's assigned to me, I'll handle it today.

Reply:
PR #1394 for this